xcell/mappers/mapper_base.py

```diff
@@ -35,3 +41,141 @@ def get_nmt_field(self, **kwargs):
+        if self.cats[mod] is None:
+            cats = []
+            for file in data_file:
+                if not os.path.isfile(file):
+                    raise ValueError(f"File {file} not found")
+                with fits.open(file) as f:
+                    cats.append(self._bin_z(Table.read(f)))
+            self.cats[mod] = vstack(cats)
+        return self.cats[mod]
+
+    def get_nz(self, dz=0):
+        if self.dndz is None:
+            cat_data = self.get_catalog(mod='data')
+            w_data = self._get_w(mod='data')
+            h, b = np.histogram(cat_data['Z'], bins=self.z_arr_dim,
+                                weights=w_data)   # , range=[0.5, 2.5])
+            self.dndz = np.array([0.5 * (b[:-1] + b[1:]), h])
+        z, nz = self.dndz
+        z_dz = z + dz
+        sel = z_dz >= 0
+        return np.array([z_dz[sel], nz[sel]])
+
+    def _get_alpha(self):
+        if self.alpha is None:
+            w_data = self._get_w(mod='data')
+            w_random = self._get_w(mod='random')
+            self.alpha = np.sum(w_data)/np.sum(w_random)
+        return self.alpha
+
+    def get_signal_map(self):
+        if self.delta_map is None:
+            self.delta_map = np.zeros(self.npix)
+            cat_data = self.get_catalog(mod='data')
+            cat_random = self.get_catalog(mod='random')
+            w_data = self._get_w(mod='data')
+            w_random = self._get_w(mod='random')
+            alpha = self._get_alpha()
+            nmap_data = get_map_from_points(cat_data, self.nside,
+                                            w=w_data)
+            nmap_random = get_map_from_points(cat_random, self.nside,
+                                              w=w_random)
+            mask = self.get_mask()
+            goodpix = mask > 0
+            self.delta_map = (nmap_data - alpha * nmap_random)
+            self.delta_map[goodpix] /= mask[goodpix]
+        return [self.delta_map]
+
+    def get_mask(self):
+        if self.mask is None:
+            if self.mask_path is None:
+                cat_random = self.get_catalog(mod='random')
+                w_random = self._get_w(mod='random')
+                alpha = self._get_alpha()
+                self.mask = get_map_from_points(cat_random,
+                                                self.nside_mask,
+                                                w=w_random)
+                self.mask *= alpha
+                # Account for different pixel areas
+                area_ratio = (self.nside_mask/self.nside)**2
+                self.mask = area_ratio * hp.ud_grade(self.mask,
+                                                     nside_out=self.nside)
+            else:
+                self.mask = hp.read_map(self.mask_path, verbose=False)
+                area_ratio = (self.nside_mask/self.nside)**2
+                self.mask = area_ratio * hp.ud_grade(self.mask,
+                                                     nside_out=self.nside)
```

Comment:
In this case you don't really need to scale by nside, I think.

Reply:
Furthermore, if we generate things from a map rather than randoms, then the generation of the delta map should also use this map, and not the randoms. I suggest we leave this option out for the time being and assume that we will always use randoms.